### ISSUE TITLE
False positive mut_mutex_lock on &mut Arc<Mutex<_>>

### ISSUE THREAD

**Issue by teor2345** [rust-lang/rust-clippy#9415]
### Summary

Clippy lints a mut_mutex_lock lint on `&mut Arc<Mutex<_>>`, which is obviously incorrect.

Applying the suggestion causes the error:
> cannot borrow data in an `Arc` as mutable

### Lint Name

mut_mutex_lock

### Reproducer

I tried this code:

```rust
use std::sync::{Arc, Mutex};

fn main() {
    let mut arc_mutex = Arc::new(Mutex::new(()));
    let arc_mutex: &mut Arc<Mutex<()>> = &mut arc_mutex;

    let _guard = arc_mutex.lock().unwrap();
}
```

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=bc5728f0494a32c7483c0d2f6fc9dc00

I saw this happen:

```
warning: calling `&mut Mutex::lock` unnecessarily locks an exclusive (mutable) reference
 --> src/main.rs:7:23
  |
7 |     let _ = arc_mutex.lock().unwrap();
  |                       ^^^^ help: change this to: `get_mut`
  |
  = note: `#[warn(clippy::mut_mutex_lock)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#mut_mutex_lock
```

I expected to see this happen:

No warnings.

### Version

```text
rustc 1.65.0-nightly (2e35f954a 2022-09-01)
binary: rustc
commit-hash: 2e35f954ada0f0c777844dc4fa66684efe90a035
commit-date: 2022-09-01
host: x86_64-unknown-linux-gnu
release: 1.65.0-nightly
LLVM version: 15.0.0
```

### Additional Labels

@rustbot label  +I-suggestion-causes-error
